libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,863 @@
+         }
+
+      private:
+         eosio::name contract_name;
+         const kv_index* idx;
+
+         uint32_t itr;
+         kv_it_stat itr_stat;
+
+         key_type key() const {
+            return idx->get_key(value());
+         }
+
+         friend kv_index;
+      };
+
+   public:
+      eosio::name name{0};
+      eosio::name table_name;
+      eosio::name contract_name;
+
+      kv_table* tbl;
+
+      kv_index() = default;
+
+      template <typename KF>
+      kv_index(KF&& kf) {
+         key_function = [=](const T& t) {
+            return make_key(std::invoke(kf, &t));
+         };
+      }
+
+      template <typename KF>
+      kv_index(eosio::name name, KF&& kf) : name{name} {
+         key_function = [=](const T& t) {
+            return make_key(std::invoke(kf, &t));
+         };
+      }
+
+      /**
+       * Search for an existing object in a table by the index, using the given key.
+       * @ingroup keyvalue
+       *
+       * @tparam K - The type of the key. This will be auto-deduced by the key param.
+       *
+       * @param key - The key to search for.
+       * @return An iterator to the found object OR the `end` iterator if the given key was not found.
+       */
+      template <typename K>
+      iterator find(const K& key) {
+         auto prefix = make_prefix(table_name, name);
+         auto t_key = table_key(prefix, make_key(key));
+
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, t_key.data(), t_key.size());
+
+         auto cmp = internal_use_do_not_use::kv_it_key_compare(itr, t_key.data(), t_key.size());
+
+         if (cmp != 0) {
+            return end();
+         }
+
+         return {contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+      }
+
+      /**
+       * Get the value for an existing object in a table by the index, using the given key.
+       * @ingroup keyvalue
+       *
+       * @tparam K - The type of the key. This will be auto-deduced by the key param.
+       *
+       * @param key - The key to search for.
+       * @return A std::optional of the value corresponding to the key.
+       */
+      template <typename K>
+      std::optional<T> get(K key) {
+         uint32_t value_size;
+
+         auto prefix = make_prefix(table_name, name);
+         auto t_key = table_key(prefix, make_key(key));
+
+         auto success = internal_use_do_not_use::kv_get(db, contract_name.value, t_key.data(), t_key.size(), value_size);
+         if (!success) {
+            return std::nullopt;
+         }
+
+         void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+         auto copy_size = internal_use_do_not_use::kv_get_data(db, 0, (char*)buffer, value_size);
+
+         datastream<const char*> ds((char*)buffer, value_size);
+
+         T val;
+         ds >> val;
```

Comment:
Will malfunction on secondary index

Reply:
performance:
* have a local `std::optional<T>` variable
* always return this variable for success or failure
* emplace then deserialize it

Reply:
That will enable return-value optimization

Reply:
The way I understand that is to have (abbreviated):
```
std::optional<T> ret_val;
...
ret_val.emplace();
ds >> ret_val;
...
return ret_val;
```

The problem I'm running into with that is datastream assumes an optional was serialized into it, however when we store a value its not an optional. So deserialization fails.

Reply:
`ds >> *ret_val`

Reply:
Thanks 👍